ambry-api/src/main/java/com.github.ambry/account/AccountBuilder.java

```diff
@@ -0,0 +1,216 @@
+
+import static com.github.ambry.account.Account.*;
+import static com.github.ambry.account.Container.*;
+
+
+/**
+ * A builder class for {@link Account}. Since {@link Account} is immutable, modifying an {@link Account} needs to
+ * build a new {@link Account} object with updated fields through this builder. An {@link Account} can be built
+ * in two ways: 1) from an existing {@link Account} object; and 2) by supplying required fields of an {@link Account}.
+ */
+public class AccountBuilder {
+  private Short id;
+  private String name;
+  private Account.AccountStatus status;
+  private short version;
+  private Map<Short, Container> idToContainerMetadataMap = new HashMap<>();
+
+  /**
+   * Constructor. This will allow building a new {@link Account} from an existing {@link Account}. The builder will
+   * include all the information including the {@link Container}s of the existing {@link Account}.
+   * @param origin The {@link Account} to build from.
+   */
+  public AccountBuilder(Account origin) {
+    if (origin == null) {
+      throw new IllegalArgumentException("origin cannot be null.");
+    }
+    this.id = origin.getId();
+    this.name = origin.getName();
+    this.status = origin.getStatus();
+    for (Container container : origin.getAllContainers()) {
+      idToContainerMetadataMap.put(container.getId(), container);
+    }
+    this.version = ACCOUNT_METADATA_VERSION_1;
+  }
+
+  /**
+   * Constructor. The builder will not include any {@link Container} information.
+   * @param id The id of the {@link Account} to build.
+   * @param name The name of the {@link Account}. Cannot be {@code null}.
+   * @param status The status of the {@link Account}. Cannot be {@code null}.
+   */
+  public AccountBuilder(short id, String name, AccountStatus status) {
+    this.id = id;
+    this.name = name;
+    this.status = status;
+    this.version = ACCOUNT_METADATA_VERSION_1;
+  }
+
+  /**
+   * Sets the id of the {@link Account} to build.
+   * @param id The id to set.
+   * @return This builder.
+   */
+  public AccountBuilder setId(short id) {
+    this.id = id;
+    return this;
+  }
+
+  /**
+   * Sets the name of the {@link Account} to build.
+   * @param name The name to set.
+   * @return This builder.
+   */
+  public AccountBuilder setName(String name) {
+    this.name = name;
+    return this;
+  }
+
+  /**
+   * Sets the status of the {@link Account} to build.
+   * @param status The id to set.
+   * @return This builder.
+   */
+  public AccountBuilder setStatus(AccountStatus status) {
+    this.status = status;
+    return this;
+  }
+
+  /**
+   * Sets a {@link Container} for the {@link Account} to build. If the builder already has a {@link Container} with
+   * the same id as new {@link Container} to set, the new {@link Container} will replace the existing {@link Container}.
+   * @param container The new {@link Container} to set.
+   * @return This builder.
+   */
+  public AccountBuilder setContainer(Container container) {
```

Comment:
I was trying some code to implement this, but then realized that this might cause some complexity to the caller. If, for example `addContainer`, a new container cannot be added if the same container id already exists, then the same logic should be applied for `updateContainer`, i.e., an exception needs to be thrown if the container does not exist. Then, the caller side may need to have many try-catch blocks and eventually leads to the same goal, add or update the container. So my feeling is a simple `addOrUpdateContainer()` might suffice, and it will not create too much confusion if it is well documented.